### ISSUE TITLE
tfjs-node "Illegal instruction: 4" error on macOS Monterey (M1)

### ISSUE THREAD

**Issue by Caundy** [tensorflow/tfjs#6381]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS 12.1 (Monterey), 2020 MacBook Air M1
- TensorFlow.js installed from (npm or script link): npm
- TensorFlow.js version: "@tensorflow/tfjs": "^3.16.0", "@tensorflow/tfjs-node": "^3.16.0"

**Describe the problem**
In an empty project, installing and requiring @tensorflow/tfjs-node results in an "Illegal instruction: 4" error thrown.

**Provide the exact sequence of commands / steps that you executed before running into the problem**
- Create an empty project (`npm init`)
- `yarn add @tensorflow/tfjs-node` in project root
- create an `index.js` file and add `const tf = require('@tensorflow/tfjs-node')` at the top of the file
- `node index.js` in project root
- observe error "Illegal instruction: 4" thrown

As reported in the closed https://github.com/tensorflow/tfjs/issues/4514 ticket, `tfjs-node` doesn't appear to work with macOS Monterey. Is there any way to make it work?

Would appreciate any tips, cheers!

**Comment by Caundy** [comment]
@pyu10055 Any chance you could take a look at the issue? :)

**Comment by pyu10055** [comment]
@Caundy I have tried to compile the latest TF with arm64 target, but it is have following errors:
```
dyld[23916]: missing symbol called
Abort trap: 6
```
I am still very confused with rosetta, I also tried to set to x86 arch with the current tfjs-node, but it also seems to fail with the same error that you have.

**Comment by MikeOBarometer** [comment]
Note that may help debugging or anyone finding this from Google.

I encountered this same issue on macOS Monterey M1 mac on v14 of node. After updating to the latest LTS version of Node `v16.15.1` and version `3.18.0` of `@tensorflow/tfjs-node`, I no longer saw this error.

**Comment by Caundy** [comment]
@MikeOBarometer thanks for reporting! Unfortunately, using these exact versions (node through nvm), after reinstalling/rebuilding packages, using Monterey 12.1, I still encounter the very same issue :/ 

@fedikan @pyu10055 Could you check whether that version combination works for you? Cheers!

**Comment by fedikan** [comment]
@Caundy @MikeOBarometer 
Same issue 
`[1]    61932 illegal hardware instruction  node index.js
`

**Comment by s314cy** [comment]
> @Caundy @MikeOBarometer Same issue on M1 pro `[1] 61932 illegal hardware instruction node index.js `

Same error here. Fixed it by installing an arm64-compatible version of node (v16+ provides that) and installing tfjs-node from there. What do you get with `node -p "process.arch"?

**Comment by pyu10055** [comment]
Hi @MikeOBarometer @Caundy 
I have tried the setup Mike mentioned, it is also working for me.
There are couple caveats you need to pay attention to:
1. Make sure you are starting your terminal without rosetta. (Which make sure the arch is in arm64)
2. use nvm to install node with arm64 binary.
3. install tfjs-node after that.

This way, your node environment arch matches with tfjs-node binding and TF download. 
And x86-64 arch that enabled by rosetta would not work with tensorflow (the illegal instruction 4 error), even though all archs match.
Please let me know if this would work for you.

**Comment by Caundy** [comment]
@pyu10055 Thank you, can confirm this solution worked, what did the trick was re-running the nvm install command before reinstalling the tfjs-node package. I had node 16 installed using nvm already but maybe did that using rosetta terminal? Anyhow, reinstallation ended up doing the trick, thank you for the help :)

**Comment by mateusgrb** [comment]
Hey folks, I also ran into the Illegal instruction error trying to run an app with Node.js 12 and "@tensorflow/tfjs-node": "^3.19.0".

After catching up with the discussion in this thread and others linked, it seems that at this time if you want to use `tfjs-node` in a M1 mac, you need to use at least Node.js 16, which is the first version that ships binaries for Apple Silicon:
https://github.com/nodejs/node/blob/main/doc/changelogs/CHANGELOG_V16.md#16.0.0
> Node.js v16.0.0 will be the first release where we ship prebuilt binaries for Apple Silicon. While we’ll be providing separate tarballs for the Intel (darwin-x64) and ARM (darwin-arm64) architectures the macOS installer (.pkg) will be shipped as a ‘fat’ (multi-architecture) binary.

In other words, upgrading my app from node 12 to node 16 is a requirement if I want to use `tfjs-node` in a M1 Mac. Is that correct? Is there any other way around it? Thanks in advance!

**Comment by MikeOBarometer** [comment]
@mateusgrb I believe node upgrades are relatively low risk but if you cannot upgrade, you could look at running your app locally in a dev container, this has some caveats I don't know off the top of my head, but I think you'll need docker in experimental mode and run it with a `--platform <target>` flag.

**Comment by basefox** [comment]
@GantMan 

try this : 

  "dependencies": {
    "@tensorflow/tfjs-node": "^3.2.0",
    "canvas": "^2.11.2",
    "express": "^4.18.2",
    "fs": "0.0.1-security",
    "http": "0.0.1-security"
  },

**Comment by kevteljeur** [comment]
If it helps anyone coming here, moving to Node.js 16 (18 in my case) from an earlier version, as well as re-installing the npm modules (remember to delete node_modules entirely) solved the problem.

**Comment by takeoto** [comment]
any news? the above suggestions don't work for me(
